source/common/ssl/context_impl.cc

```diff
@@ -370,6 +381,142 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
+      std::copy_n(src_key.begin() + pos, dst_key.hmac_key.size(), dst_key.hmac_key.begin());
+      pos += dst_key.hmac_key.size();
+      std::copy_n(src_key.begin() + pos, dst_key.aes_key.size(), dst_key.aes_key.begin());
+      pos += dst_key.aes_key.size();
+      ASSERT(src_key.begin() + pos == src_key.end());
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(
+        ctx_.get(),
+        [](SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+           int encrypt) -> int {
+          ContextImpl* context_impl = static_cast<ContextImpl*>(
+              SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), sslContextIndex()));
+          return dynamic_cast<ServerContextImpl*>(context_impl)
+              ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+        });
+  }
+
+  // Hash all the settings that affect whether the server will allow/accept
+  // the client connection.  This ensures that the client is always validated against
+  // the correct settings, even if session resumption across different listeners
+  // is enabled.
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  EVP_MD_CTX md;
+  int rc = EVP_DigestInit(&md, EVP_sha256());
+  RELEASE_ASSERT(rc == 1);
+
+  // Include "envoy" so that if all the other verify-settings are unset
+  // we have a deterministic value.
+  const char* initial_context = "envoy";
+  rc = EVP_DigestUpdate(&md, initial_context, strlen(initial_context));
+  RELEASE_ASSERT(rc == 1);
+
+  if (ca_cert_ != nullptr) {
+    rc = X509_digest(ca_cert_.get(), EVP_sha256(), session_context_buf, &session_context_len);
+    RELEASE_ASSERT(rc == 1 && session_context_len == SHA256_DIGEST_LENGTH);
+    rc = EVP_DigestUpdate(&md, session_context_buf, session_context_len);
+    RELEASE_ASSERT(rc == 1);
+
+    // verify_subject_alt_name_list_ can only be set with a ca_cert
+    for (const std::string& name : verify_subject_alt_name_list_) {
+      rc = EVP_DigestUpdate(&md, name.data(), name.size());
+      RELEASE_ASSERT(rc == 1);
+    }
+
+    // verify_certificate_hash_ can only be set with a ca_cert
+    rc = EVP_DigestUpdate(&md, verify_certificate_hash_.data(),
+                          verify_certificate_hash_.size() *
+                              sizeof(decltype(verify_certificate_hash_)::value_type));
+    RELEASE_ASSERT(rc == 1);
+  }
+
+  rc = EVP_DigestFinal(&md, session_context_buf, &session_context_len);
+  RELEASE_ASSERT(rc == 1);
+  rc = SSL_CTX_set_session_id_context(ctx_.get(), session_context_buf, session_context_len);
+  RELEASE_ASSERT(rc == 1);
+}
+
+int ServerContextImpl::sessionTicketProcess(SSL*, uint8_t* key_name, uint8_t* iv,
+                                            EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx, int encrypt) {
+  const EVP_MD* hmac = EVP_sha256();
+  const EVP_CIPHER* cipher = EVP_aes_256_cbc();
+
+  if (encrypt == 1) {
+    // Encrypt
+    if (session_ticket_keys_.empty()) {
```

Comment:
Sorry, I assumed this would fallback to "don't send SessionTicket", but that's not the case and we can't have an empty `session_ticket_keys_` list after the callback is configured. Please revert this to `RELEASE_ASSERT()` that you had before.

We'll need to enforce this in SDS (feel free to add TODO).

Reply:
reverted